[Zoonmy] step-1 index.html 응답

pjm2571 opened the pull request:
## 구현 내용
- 경로를 나타내기 위한 ROUTE Enum 구현
- Content-Type을 나타내기 위한 Content-type enum 구현
- concurrent 패키지의 newFixedThreadPool을 사용
- .html, .css, .ico, .svg 파일의 로드 구현

## 고민 사항
- concurrent 패키지의 newFixedThreadPool을 사용하면서 학습한 키워드 중에 submit(), execute()가 있었는데 아직까지는 어떠한 상황에서 어떠한 메소드를 써야할 지 감이 잘 잡히지 않아, 조금 더 학습이 필요한 것 같습니다.
- dataoutputstream-header를 조작할 때, content-type에 따라 타입이 결정된다는 것을 학습하였고, 만약 .html이 아닌 /index와 같이 경로로 주어지게 된다면 어떻게 구현해야 할 지 고민해보게 되었습니다.
- 제 기존 코드의 경우에는 static에 있는 파일들만 가져오도록 구현하였기 때문에 리팩토링이 필요할 것 같습니다.
- 1주차 JK의 리뷰인 'Enum의 getter 사용을 지양하고 Enum 그 자체로 동작하도록 구현' 을 시도해보려고 노력해보았습니다. path를 매개변수로 받아 절대경로를 return하는 enum과 type을 매개변수로 받아 full-content-type으로 리턴하도록 구현해보았는데, 이렇게 구현하는 것이 Enum의 getter를 지양하고 enum 타입 그대로 사용하는 것인지 고민이 많이 되었습니다..

## 기타

pjm2571 committed:
feat: static 정적 파일 응답 구현
- localhost:8080/파일.html을 입력할 때, 정적인 파일 로드
- css는 동작하지 않는다.

pjm2571 committed:
feat: concurrent 패키지로 스레드 구조 변경
- ExecutorService.newFixedThreadPool을 통해 크기가 100인 스레드 풀 생성
- executor.execute()를 통해 클라이언트 socket의 스레드 생성

pjm2571 committed:
feat: HTTP Request 내용 출력 구현
- request 부분
- header 부분
- body 부분 : index.html 내용
들이 출력되도록 구현

pjm2571 committed:
refactor: thread pool size 상수화
- static int -> static final int 로 수정

pjm2571 committed:
feat: StringUtils 클래스 구현
- 문자열을 받아서 공백을 기준으로 토크나이즈
- 토큰화한것들 중 1번째 값을 반환

pjm2571 committed:
feat: Path와 Type을 파싱해주는 StringUtils 클래스 구현
- getPath() : path를 리턴
- getType() : type을 리턴

pjm2571 committed:
feat: ROUTE를 지정해주는 enum 구현
- path를 매개변수로 받아서, ROUTE를 설정하여 return

pjm2571 committed:
feat: StringUtils 클래스를 검증하는 테스트 코드 구현
- 올바른 request line 입력 시, path 추출
- 올바르지 않은 request line 입력 시, 예외 처리

- 메소드 확장자 타입 체크

pjm2571 committed:
feat: path를 매개변수로 Content-Type을 리턴하는 Enum 구현
- path를 입력받아서 content-type을 리턴

pjm2571 committed:
feat: content-type을 검증하는 테스트 코드 구현
- html, css, svg, ico 파일에 대한 정상 반환 테스트
- html 과 css 파일에 대한 miss match 테스트

pjm2571 committed:
feat: 정적인 index.html 응답, HTTP Request 출력 구현
- request Line이 /index.html 로 주어질 때
	- static/index.html을 매핑
	- static/img 파일을 매핑
	- static.css 파일들을 매핑
		- Content-Type 분리를 통해 해결
- logger.debug를 활용하여 각각의 메소드 라인마다 message를 붙여 구현

pjm2571 committed:
docs: REAMDE.md 파일, week1.md 파일 작성
- 주간 학습 계획 작성
	- Day - 1 작성
- README.md 파일 작성

honux77 commented on docs/week1.md:
````diff
@@ -0,0 +1,91 @@
+- ```Executor```인터페이스를 구현한 클래스이다.
+
+- ExecutorService의 newFixedThreadPool을 이용하여 미션을 구현해본다.
+  - 크기를 100으로 한정하여 먼저 생성해본다.
````
왜 100일까요? 관련 내용을 조금 더 공부해 보면 좋겠습니다.

pjm2571 replied:
네! 알겠습니다! concurrent 패키지 관련하여 학습하고, ```docs/refactor.md``` 파일로 리뷰받은 것들을 정리하고 학습해서 다시 커밋 하겠습니다

honux77 commented on src/main/java/webserver/CONTENT_TYPE.java:
```diff
@@ -0,0 +1,20 @@
+package webserver;
+
+public enum CONTENT_TYPE {
```
enum이라도 클래스의 이름을 대문자로 사용하는 일은 별로 없는 것 같아요. 사용 예가 있는지 확인 좀 해 주실래요?

honux77 replied:
컨텐츠 타입은 충분히 enum이 될 수 있는데 지금 단계에서는 상수 문자열로 하는 것보다 복잡도만 증가시키는 것 같은 느낌이 들어요.

pjm2571 replied:
> enum이라도 클래스의 이름을 대문자로 사용하는 일은 별로 없는 것 같아요. 사용 예가 있는지 확인 좀 해 주실래요?

네! ContentType 으로 클래스의 이름 수정하였습니다. 리팩토링을 해야겠다고 생각했었는데, 구현을 하다 보니 놓쳐서 리팩토링하지 못한 부분이었습니다. 앞으로 주의해서 구현하겠습니다.

honux77 commented on src/main/java/webserver/ROUTE.java:
STATIC 경로 외에 나머지는 필요 없어 보입니다. static 경로 내의 모든 파일의 경로는 URL의 상대 경로와 같거든요.

honux77 commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -24,19 +29,53 @@ public void run() {
+     * requestLine : httpRequestLine
+     * -> logger.debug 메소드를 활용해서 입력받은 requestLine을 디버그해준다.
+     */
+    private void loggerDebug(String requestMessage, String requestLine) {
+        logger.debug(requestMessage + " : {}", requestLine);
+    }
```
없어도 될 것 같은 느낌인데요?

honux77 commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -1,17 +1,22 @@
-import java.io.OutputStream;
+import java.io.*;
 import java.net.Socket;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
```
nio 는 사용 금지입니다.

honux77 reviewed (changes requested):
nio 사용한 부분 제거해 주세요.

pjm2571 committed:
refactor: concurrent의 ThreadPool 수정
- newFixedThreadPool을 ThreadPoolExecutor로 정의하여 사용
	- 최소 코어 스레드 : 3
	- 최대 스레드 : 100
	- 노는 시간 : 120초
로 설정하여 스레드 풀을 관리한다.

pjm2571 committed:
refactor: ContentType Enum 수정
- CONTENT_TYPE -> ContentType으로 Enum 클래스 명 수정
- ContentType이 복잡한 String 구현 대신, 단순히 type 값만 넘겨주도록 설정

pjm2571 committed:
refactor: Route Enum 수정
- ROUTE -> Route 로 클래스 이름 수정
- Route 내부의 값을 STATIC, TEMPLATE으로 두어 상대 경로로 동작하도록 설정

pjm2571 committed:
refactor: loggerDebug 함수 삭제
- loggerDebug 함수를 삭제
- 불필요한 상수 REQUEST, HEADER 삭제

pjm2571 committed:
refactor: java.nio 제거, java.io 패키지 사용
- FileInputStream과 BufferedInputStream을 사용하여 파일을 읽어오게 구현

honux77 reviewed (approved):
빠른 반영 좋아요. 수고하셨습니다.
